### DISCUSSION TITLE
Custom fields, tools, and resource tools feedback

### DISCUSSION THREAD

**Discussion by adrianthedev** [avo-hq/avo#836] [Feature feedback]
Let's use this space to provide feedback for anything custom-content-related. This includes custom fields, custom tools, custom asset pipelines, and general Avo customization.

Post [an issue](https://github.com/avo-hq/avo/issues/new/choose) if something is broken.

**Comment by dhnaranjo** [comment]
I'm using a custom field to render a preview of a card in the user facing app. It uses several attributes from the object so there isn't a single method to get its data. 
```
class ItemResource < Avo::BaseResource
  # ...
    field :preview, # <-- This method does not exist
      as: :item_preview, 
      only_on: :show
  # ...
end
```

That's fine, I just grab `@field.model` in the view and get to it, except that the default behavior for a field without a value is to replace the fields content with a... is that an emdash? Anyway. To get around this I:

```
class ItemPreviewField < Avo::Fields::BaseField
  def initialize(name, **args, &block)
    super(name, **args, &block)
  end

  def value = true
end
```

Now it has a value and renders as expected. Works, but not obvious. Seems like there's opportunity to... something. I dunno.

**Reply by adrianthedev** [reply]
Yes, there is a way. In the edit/show/index components, pass this attribute to the `edit_field_wrapper` `dash_if_blank: false`. Let me know if that fixes it.

**Reply by dhnaranjo** [reply]
Oh yeah that works just right. It feels like something I'd rather set / expect to find in the `ItemPreviewField` class definition.

**Reply by adrianthedev** [reply]
Is there such a class in Avo?😅
For me it was easier to pass it to the wrapper form each field.

**Reply by dhnaranjo** [reply]
That was the name of the custom field class I created n referenced in my initial comment, pardon for being unclear. So I guess I would say that I'd prefer to set it in the subclass of `Avo::Fields::BaseField`, as opposed to my hacked `def value = true` bit.

**Reply by adrianthedev** [reply]
Ok. I understand now.

I chose the template way because it gives you more flexibility per view. If you only need to display the `em dash` on Index, you can do that from the view. If you had to do it in the class you'd have to use an if statement (or ternary) to achieve that.

**Comment by brandonzylstra** [comment]
https://docs.avohq.io/2.0/customization.html#home-path says 

> Required v2.8.0+

At first I thought that was referring to the version of Avo, but the [Roadmap](https://avohq.io/roadmap) lists 2.0 as the current version, so apparently this is referring to the Ruby version?   Nope, there's no such thing as 2.8.0 according to https://www.ruby-lang.org/en/

What is required to be at version 2.8.0+?

**Reply by adrianthedev** [reply]
I think I see the confusion. Correct me if I'm wrong, please. The roadmap says `v2.0` instead of `v2.x` "series two" of the package. I tweaked that messaging.

The docs refer to the Avo version required for that specific option. I tweaked that in the docs too.

**Reply by brandonzylstra** [reply]
Ahhhh, I understand now.  Thanks!

**Comment by brandonzylstra** [comment]
@dhnaranjo wrote 

> I'm using a custom field to render a preview of a card in the user facing app. 

This touches on something I'm currently trying to figure out about Avo, so I hope it's okay to ask this here....  and apologies in advance if this is really basic... I'm currently brand new to Avo and just at the stage of adding it to my app.  

Is there any documentation explaining how to embed elements created by Avo into an existing app template?  Is it possible to do this, or is it currently necessary for Avo to generate the entire page?

**Reply by adrianthedev** [reply]
> embed elements created by Avo into an existing app template

Do you mean, for example, to embed the table that lists all the records in a non-Avo page? Like a page that renders from one of your controllers and on your erb's?

That is technically possible because those are View components. The downside is that you have to shape the data that the view components ingest in the same way Avo does, which can be a bit painful at times.

What I'm trying to say is that those components weren't **made** to support that, but they can. I'm exploring that path a bit in https://github.com/avo-hq/avo/pull/1138, but supporting **all elements**  is not going to be a plug-and-play thing, but a fix-one-at-a-time thing.

So it's best to use them inside Avo for now.

**Reply by dhnaranjo** [reply]
> @dhnaranjo wrote 
> 
> 
> 
> > I'm using a custom field to render a preview of a card in the user facing app. 
> 
> 
> 
> This touches on something I'm currently trying to figure out about Avo, so I hope it's okay to ask this here....  and apologies in advance if this is really basic... I'm currently brand new to Avo and just at the stage of adding it to my app.  
> 
> 
> 
> Is there any documentation explaining how to embed elements created by Avo into an existing app template?  Is it possible to do this, or is it currently necessary for Avo to generate the entire page? 

I think you understood this, but just to make sure, I was using a primary app component in Avo, not the other way round. It worked fine.

**Reply by adrianthedev** [reply]
Hmm, not sure I understand. What is _a primary app component_?

**Reply by dhnaranjo** [reply]
> Hmm, not sure I understand. What is _a primary app component_?

My goof, I should have replied to the question you replied to, not your reply. Mobile app woes. 

What I meant was, dude wants to use Avo components in his app, while my situation was using my apps components in Avo.
